snapcraft/parts/plugins/flutter.py

```diff
@@ -0,0 +1,64 @@
+from typing import List, Dict, Set, cast, Literal, Any
+
+from craft_parts import plugins
+
+FLUTTER_REPO = "https://github.com/flutter/flutter.git "
+
+class FlutterPluginProperties(plugins.PluginProperties, plugins.PluginModel):
+    flutter_branch: Literal["stable", "master", "dev"] = "stable"
+    flutter_target: str = "lib/main.dart"
+
+    @classmethod
+    def unmarshal(cls, data: Dict[str, Any]) -> "CondaPluginProperties":
+        """Populate class attributes from the part specification.
+
+        :param data: A dictionary containing part properties.
+
+        :return: The populated plugin properties data object.
+
+        :raise pydantic.ValidationError: If validation fails.
+        """
+        plugin_data = plugins.extract_plugin_properties(
+            data,
+            plugin_name="flutter",
+        )
+        return cls(**plugin_data)
+
+
+class FlutterPlugin(plugins.Plugin):
+
+    properties_class = FlutterPluginProperties
+
+    def get_build_snaps(self) -> Set[str]:
+        return {}
+
+    def get_build_packages(self) -> Set[str]:
+        return {
+            "clang",
+            "git",
+            "cmake",
+            "ninja-build",
+            "unzip",
+        }
+
+    def get_build_environment(self) -> Dict[str, str]:
+        return {
+            "PATH": "$CRAFT_PART_BUILD/flutter-distro/bin:$PATH",
+        }
+
+    def _get_setup_flutter(self, options):
+        return [
+            f"[ -d flutter-distro ] || git clone -b {options.flutter_branch} {FLUTTER_REPO} flutter-distro",
+            "flutter doctor",
+            "flutter pub get",
```

Comment:
I'd suggest running `flutter precache --linux` here to download the Flutter binary artifacts for Linux at the setup phase (or else it will be done at the build phase)

```
Usage: flutter precache [arguments]
-h, --help              Print this usage information.
-a, --all-platforms     Precache artifacts for all host platforms.
-f, --force             Force re-downloading of artifacts.
    --[no-]ios          Precache artifacts for iOS development.
    --[no-]web          Precache artifacts for web development.
    --[no-]linux        Precache artifacts for Linux desktop development.
    --[no-]windows      Precache artifacts for Windows desktop development.
    --[no-]macos        Precache artifacts for macOS desktop development.
    --[no-]fuchsia      Precache artifacts for Fuchsia development.
    --[no-]universal    Precache artifacts required for any development platform.
                        (defaults to on)
```

Reply:
good idea